### PR TITLE
trace: print SOF version numbers at boot also in Zephyr builds

### PR DESCRIPTION
**Pull request by kv2019i**
The log print for SOF build and ABI version numbers at boot time, is only printed in dma_trace_enable(). It is done here to ensure the log is not emitted until logging infrastructure is in place and the log print is not lost.

The information is very useful, so similar log should be emitted also in SOF Zephyr builds, where the Zephyr logging subsystem is used instead of SOF's own trace infra.

As Zephyr logging subsystem is available early, this code can now be added directly to trace_init(), which is called from sof_main().

Signed-off-by: Kai Vehmanen <kai.vehmanen@linux.intel.com>

### CODE REVIEW

File: src/trace/trace.c [hunk 1]
```diff
@@ -497,6 +503,28 @@ void trace_off(void)
...
+	 */
+#ifdef __ZEPHYR__
+	LOG_INF("FW ABI 0x%x DBG ABI 0x%x tags SOF:" SOF_GIT_TAG " zephyr:" \
+		META_QUOTE(BUILD_VERSION) " src hash 0x%08x (ref hash " \
+		META_QUOTE(SOF_SRC_HASH) ")",
+		SOF_ABI_VERSION, SOF_ABI_DBG_VERSION, SOF_SRC_HASH);
+#endif
+}
+
 void trace_init(struct sof *sof)
```
Comment (abonislawski) [inline src/trace/trace.c:528]: CONFIG_TRACE=n
CONFIG_LOG=y
CONFIG_ZEPHYR_LOG=y

is this scenario valid? Because it wont print in such case
Not sure if we need to compile trace files with zephyr logs?

Comment (kv2019i) [reply src/trace/trace.c:528]: @abonislawski Currently if you set CONFIG_TRACE=n , you'll only get Zephyr side logs, SOF side logs will be omitted via include/sof/trace/trace.h . 

It kind of makes sense not to print the SOF version.

Combination:
CONFIG_TRACE=n
CONFIG_ZEPHYR_LOG=y

is not correct though as ZEPHYR_LOG will have (currently) impact only if CONFIG_TRACE is selected.

**Comment by kv2019i** [comment]
@marc-hb wrote:
> Could you add a prefix for consistency with the existing `SHM:` and `DMA:`?

Considered this, but I left out with following reasoning:
 - the prefix will be different anyways this is output via other channel (logging subsystem will format the logs)
 - what would be the prefix? the log may go out via multiple backends and I don't really no in trace.c (it could be SHM or DMA)

I can add back if good proposal what to use as prefix..

> Have you considered moving SOF_BANNER_COMMON to some _version.h file?

I actually started with this. I was thinking to strip the SRC_HASH and Zephyr version (as it is redundant) and thus defining these locally seemed like the less complex approach. It's already quite complicated with the many generated version headers (for SOF and Zephyr), plus the SRC_HASH generation in cmake rules.

But, but, I actually added all the info back. Dictionary implementation is somewhat different, but probably the SRC hash (including the duplicated format and parameter copies) are probably still useful for debug. And just for convenience, having all version numbers on a single line is just nice, so added Zephyr as well. 

So long story short, yes, probably sharing the SOF_BANNER_COMMON somewhere makes sense. Let me see if I can find a good place for it.

File: src/trace/trace.c (hunk 1 above)
Comment (abonislawski) [reply src/trace/trace.c:528]: yeap but actually this is good because there is no point in compiling all trace stuff if we want to call zephyr LOG_ anyway.
So can you move this somewhere else print this without trace too?

**Comment by marc-hb** [comment]
> the prefix will be different anyways this is output via other channel (logging subsystem will format the logs)

OK, as long as it never looks exactly the same and is recognizable then we're good.

The "requirement" / worst use case here is to quickly and unambiguously identify the origin of incomplete logs shared in an incomplete bug poorly filed by someone who doesn't know anything and just want their audio to work. You get the idea. If one or two extra characters can solve this problem once for all then we would forever regret not adding them but you convinced me this will never be a problem... right? Promise? :-)

> what would be the prefix? the log may go out via multiple backends and I don't really no in trace.c (it could be SHM or DMA)

I thought about something like "ZEP" for Zephyr but no need based on the above.

> So long story short, yes, probably sharing the SOF_BANNER_COMMON somewhere makes sense. Let me see if I can find a good place for it.

Thanks, appreciated. As you can tell from #3890 I think the SRC_HASH design is flawed but it was much better than nothing at all. It was also designed for the sof-logger which is being deprecated but even without sof-logger it cannot hurt.

So as long as the line's length stays reasonable (and I don't see why it would grow any further) and as long as there's a prefix or something to disambiguate, there is value in:
- not confusing people browsing the source with the same line duplicated in two different places
- output consistency. There's at least one function in sof-test already searching for this line.

File: src/trace/trace.c (hunk 1 above)
Comment (marc-hb) [reply src/trace/trace.c:528]: > I was assuming CONFIG_TRACE ifdef covers all the trace macros, but it doesn't.

The best recipe to debug pre-processing is the same as the one for debugging build system issues: **break it** on purpose.

So you're not sure whether some code / macro is used in some specific CONFIG_uration? Easy: add a syntax error in the source. If it still compiles then you know for sure it's not used. This is literally a million times faster than reading the source or trying to understand the output of `gcc -E`.

Even better: when it fails to compile the compiler shows you the actual stack of macros, something literally impossible to figure out by just reading the source past some macro nesting and Kconfig complexity.

It's a bit like test-driven development where the only way to "test the test" is to (re-)introduce a bug.

Error handling in compilers is seriously underrated. There's a basic but very informative and super easy to use source code analyzer hidden in compiler error messages.

**Comment by kv2019i** [comment]
V2 update:
- moved to init so this is printed even if CONFIG_TRACE=n
- no common banner (the term "ldc" has is specific to old dictionary so I decided it's easier to have a separate format definition in init.c)

File: src/init/init.c [hunk 2]
```diff
@@ -31,6 +31,13 @@
 #if CONFIG_IPC_MAJOR_4
 #include <ipc4/fw_reg.h>
 #endif
+#include <user/abi_dbg.h>
```
Comment (abonislawski) [inline src/init/init.c:34]: we need this (and the rest) without ZEPHYR_LOG?

Comment (kv2019i) [reply src/init/init.c:34]: Ack @abonislawski , will fix in V3.

**Comment by kv2019i** [comment]
V3 update:
- added preprocessor  guards to version related headers only needed with zephyr logging

**Review by abonislawski — APPROVED** [review]

**Review by marc-hb — COMMENTED** [review]
https://sof-ci.01.org/sof-pr-viewer/#/build/PR6378/build10540898 is red but QB had a serious outage yesterday; maybe wait and force push again in a day or two.

@kv2019i can you explain why cavstool.py shows some Zephyr logs in https://sof-ci.01.org/sofpr/PR6378/build1816/devicetest/?model=ADLP_RVP_SDW_ZEPHYR&testcase=check-sof-logger -> logger.etrace but not this new one? Sorry I'm getting lost in all the too many logging possibilities

**Review by lyakh — APPROVED** [review]

**Comment by kv2019i** [comment]
One known checkpatch warning that I couldn't figure out how to avoid:
"WARNING: Avoid unnecessary line continuations
#62: FILE: src/init/init.c:212:
+	LOG_INF("FW ABI 0x%x DBG ABI 0x%x tags SOF:" SOF_GIT_TAG " zephyr:" \"

... I think it's much more readable this way, so I'm sticking to this one.

As there are no other failing tests, I'll proceed to merge this as having the version included in logs is urgent for CI result "trackability". Let's polish the output in later PRs if needed.
